### DISCUSSION TITLE
Help with "nontrival" Snakefile and showyourwork.yml

### DISCUSSION THREAD

**Discussion by stefanbringuier** [rodluger/showyourwork#414] [Q&A] [answered]
I'm developing a Snakemake workflow for a project that requires a finely tuned orchestration of simulation data generation based on various models and configurations. 

### Workflow Requirements:

1. The script needs to execute in specific conda environments for each model/configuration pair. 
    - This works thanks to #413. 
3.  A global database file is crucial and must be initialized at the start of the workflow to prevent contamination from previous runs. 
    - Have gotten it to work, but breaks easily.
5. The database should be cached once all simulations have been completed. 
    - Have gotten all the rules to run but not with caching. I have Zenodo properly configured.

I think I've come up with a solution shown below in the `Snakefile`, but I'm not sure this is the best approach and I have some issues as mentioned below. Improvements are greatly appreciated!

```snakemake
MODELS = ["ModelA","ModelB","ModelC"]
CONFIGURATIONS = ["Config1","Config2","Config3"] 
DATABASE = "Database.json"

def get_env_file(model):
    return f"env/{model.lower()}.yml"
 
# Rule to create the database
rule create_db:
    input:
        "src/scripts/CreateMetadata.py",
        "src/scripts/CreateDatabase.py"
    output:
        "src/data/" + DATABASE
    conda:
        "env/base.yml"
    params:
        dbname=DATABASE
    shell:
        "python src/scripts/CreateDatabase.py {params.dbname}"

# Rule to add data to the database
rule add_data_to_db:
    input:
        "src/data/" + DATABASE,
        "src/scripts/PopulateData.py"
    output:
        touch("src/data/COMPLETED_TASKS/{configuration}_{model}.data.done")
    conda:
        lambda wildcards: get_env_file(wildcards.model)
    params:
        dbname=DATABASE,
        configuration=lambda wildcards: wildcards.configuration,
        model=lambda wildcards: wildcards.model
    shell:
        "python src/scripts/PopulateData.py --dbname {params.dbname} --configuration {params.configuration} --model {params.model}"

# Rule to aggregate data into a single database
rule aggregate_db:
    input:
        expand("src/data/COMPLETED_TASKS/{configuration}_{model}.data.done", configuration=CONFIGURATIONS, model=MODELS)
    output:
        touch("src/data/COMPLETED_TASKS/db.aggregate.done")

# Rule to trigger the full pipeline & cache
rule run_all_calculations:
    input:
        "src/data/COMPLETED_TASKS/db.aggregate.done"
    output:
        directory("src/data")
    cache:
        True

# Rule to plot data comparison
rule plot_db_comparison:
    input:
        "src/data"
    output:
        "src/tex/figures/ModelComparison.png"
    conda:
        "env/base.yml"
    script:
        "src/scripts/PlotData.py"
```
### Challenge and Questions:

I have in my `showyourwork.yml` file, the following to capture the folder dependency for the plotting rule:

```yml
dependencies:
   src/script/PlotData.py
     - src/data/
```

But I get an error when running `showyourwork build`

```shell
Traceback (most recent call last):
  File "/home/user/project/.snakemake/scripts/tmpcveay1ez.preprocess.py", line 495, in <module>
    config["tree"] = get_json_tree(snakemake.input[0])
  File "/home/user/project/.snakemake/scripts/tmpcveay1ez.preprocess.py", line 368, in get_json_tree
    dependencies = config["dependencies"].get(script, [])
AttributeError: 'str' object has no attribute 'get'
Error in rule syw__preprocess:
    jobid: 1
    output: .showyourwork/config.json

```

1. Does my `Snakefile` make sense and is this how I should configure the rules?
2. What am I doing wrong in the `showyourwork.yml` file?

**Comment by dfm** [comment, marked as answer]
This all looks about right to me, and I'm not quite sure why you're getting that error. But, I don't believe that you need that `dependencies` entry in your `showyourwork.yml` config, since it's already specified by the `Snakefile`. I'd recommend removing that config entry, then deleting your `.showyourwork` and `.snakemake` directories then running again. Let me know what you find!

**Reply by stefanbringuier** [reply]
Thanks for taking the time to go over my `Snakefile`! Let me take your suggestions and give it a try, the one reason I added the `dependencies` part was I was under the impression this is needed to assert caching on Zenodo sandbox, but I think I was confused. I think I understand that the `dependencies` section in `showyourwork.yml` is used to achieve the dependency trigger of a rule, as opposed to how I'm doing it in the `Snakefile` by creating empty checkpoint files. Although maybe I still misunderstand. 

The thing that I don't get is `User authentication for 10.XXXX/zenodo.XXXXXXXX is valid.` for Zenodo, but when I go check the Sandbox nothing is there.

**Reply by stefanbringuier** [reply]
So I managed to construct a `Snakefile` that behaves as anticipated.

I still have issues with Zenodo Sandbox, in that it appears I have a valid Zenodo Sandbox entry/DOI, and it authenticates, but when I check after I've run `showyourwork build` , there is nothing updated on the Zenodo Sandbox. I have `cache: True` placed in the `aggregate_db` rule. I've tried other places but that gives an issue.

A related issue #418  w.r.t. github actions showed up for this `Snakefile` that uses multiple conda environments.

**Reply by stefanbringuier** [reply]
Marking this as answered at least for the `Snakemake` file working. Still have issue with caching and Zenodo for this workflow.
